docs/language/learn-ql/writing-queries/debugging-queries.rst

```diff
@@ -0,0 +1,92 @@
+Query writing: common performance issues
+========================================
+
+This topic offers some simple tips on how to avoid commons problems that can affect the performance of your queries.
+Before reading the tips below, it is worth reiterating a few important points about CodeQL and the QL language:
+
+- In CodeQL `predicates <https://help.semmle.com/QL/ql-handbook/predicates.html>`__ and `classes <https://help.semmle.com/QL/ql-handbook/types.html#classes>`__ are all just database `relations <https://en.wikipedia.org/wiki/Relation_(database)>`__---that is, sets of tuples in a table. Large predicates generate tables with large numbers of rows, and are therefore expensive to compute.
+- The QL language is implemented using standard database operations and `relational algebra <https://en.wikipedia.org/wiki/Relational_algebra>`__ (such as join, projection, union, etc.). For further information about query languages and databases, see :doc:`About QL <../about-ql>`.
+- Queries is evaluated *bottom-up*, which means that a predicate is not evaluated until **all** of the predicates that it depends on are evaluated. For more information on query evaluation, see `Evaluation of QL programs <https://help.semmle.com/QL/ql-handbook/evaluation.html>`__ in the QL handbook. 
+
+Performance tips
+----------------
+
+Follow the guidelines below to ensure that you don't get get tripped up by the most common CodeQL performance pitfalls.
+
+Eliminate cartesian products
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The performance of a predicate can often be judged by considering roughly how many results it has. 
+One way of creating badly performing predicates is by using two variables without relating them in any way, or only relating them using a negation.
+This leads to computing the `Cartesian product <https://en.wikipedia.org/wiki/Cartesian_product>`__ between the sets of possible values for each variable, potentially generating a huge table of results.
+
+This can occur whenever you inadvertently fail to specify restrictions on your variables. 
+
+For instance, in the following case none of the parameters are related in the example predicate ``methodAndAClass``, and therefore the results are unrestricted::
+
+   // BAD! Cross-product
+   predicate methodAndAClass(Method m, Class c) {
+   	  any()
+   }
```

Comment:
This example is a bit too contrived. How about this?

---

For instance, consider the following predicate that checks whether a Java method `m` may access a field `f`:

```
predicate mayAccess(Method m, Field f) {
  f.getAnAccess().getEnclosingCallable() = m
  or
  not exists(m.getBody())
}
```

The predicate holds if `m` contains an access to `f`, but also conservatively assumes that methods without bodies (for example native methods) may access any field at all.

What this means, however, is that if `m` is a native method, then the table computed by `mayAccess` contains a row `m, f` for _all_ fields `f` in the code base, resulting in a potentially very large table.